lib/ecto/changeset.ex

```diff
@@ -556,14 +556,24 @@ defmodule Ecto.Changeset do
     case params do
       %{^param_key => value} ->
         value = if value in empty_values, do: Map.get(defaults, key), else: value
-        case Ecto.Type.cast(type, value) do
+        case Ecto.Type.cast(type, value, current) do
           {:ok, value} ->
             if Ecto.Type.equal?(type, current, value) do
               :missing
             else
               {:ok, value, valid?}
             end
 
+          {:ok, value, valid2?} ->
+            if Ecto.Type.equal?(type, current, value) do
```

Comment:
I was ideally thinking we get rid of `equal?` and have it return `:ignore` if it is the same but that would mean adding a branch to `Ecto.Changeset` only to handle parameterized types... which is not ideal. 🤔